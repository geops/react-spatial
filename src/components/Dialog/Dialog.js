import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { MdClose } from 'react-icons/md';
import Draggable from 'react-draggable';
import Header from '../Header';
import Button from '../Button';

const propTypes = {
  /**
   * Children content of the Dialog.
   */
  children: PropTypes.element,

  /**
   * CSS class of the dialog.
   */
  className: PropTypes.string,

  /**
   * CSS class of the dialog's children.
   */
  classNameChildren: PropTypes.string,

  /**
   * CSS class of the dialog's header.
   */
  classNameHeader: PropTypes.string,

  /**
   * CSS class of the dialog close button.
   */
  classNameCloseBt: PropTypes.string,

  /**
   * Specifies a selector to be used to prevent drag initialization.
   * Pass to 'cancel' props of sub-component 'react-draggable'.
   * https://github.com/mzabriskie/react-draggable
   * (Only available if isDraggable is true)
   */
  cancelDraggable: PropTypes.string,

  /**
   * Function triggered on closing dialog.
   */
  onClose: PropTypes.func.isRequired,

  /**
   * If set to true, allow user to drag the dialog.
   */
  isDraggable: PropTypes.bool,

  /**
   * If set to true, blocks the application, and disable dragging action.
   */
  isModal: PropTypes.bool,

  /**
   * If set to true, dialog is displayed.
   */
  isOpen: PropTypes.bool,

  /**
   * Dialog title.
   */
  title: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),

  /**
   * Default Dialog position (Only available if isDraggable is true).
   */
  position: PropTypes.shape({
    x: PropTypes.number,
    y: PropTypes.number,
  }),

  /**
   * Function triggered on drag stop (Only available if isDraggable is true).
   */
  onDragStop: PropTypes.func,
};

const defaultProps = {
  children: undefined,
  className: 'tm-dialog-container',
  classNameChildren: undefined,
  classNameHeader: 'tm-dialog-header',
  classNameCloseBt: 'tm-button tm-dialog-close-bt',
  cancelDraggable: undefined,
  isDraggable: false,
  isModal: false,
  isOpen: false,
  title: undefined,
  position: null,
  onDragStop: () => {},
};

/**
 * This component creates a Dialog window.
 */
class Dialog extends Component {
  renderDialogTitle() {
    const {
      title,
      onClose,
      classNameHeader,
      classNameCloseBt,
      isDraggable,
    } = this.props;
    return (
      <Header
        className={`${classNameHeader} ${
          isDraggable ? 'tm-dialog-draggable' : ''
        }`}
      >
        {title}
        <Button className={classNameCloseBt} onClick={() => onClose()}>
          <MdClose focusable={false} />
        </Button>
      </Header>
    );
  }

  renderDialog() {
    const {
      className,
      classNameChildren,
      children,
      isModal,
      isOpen,
    } = this.props;
    return (
      <div
        className={isModal ? 'tm-modal' : 'tm-dialog'}
        style={{ display: isOpen ? 'block' : 'none' }}
      >
        <div className={className}>
          {this.renderDialogTitle()}
          <div className={classNameChildren}>{children}</div>
        </div>
      </div>
    );
  }

  render() {
    const {
      isModal,
      isOpen,
      isDraggable,
      onDragStop,
      position,
      cancelDraggable,
    } = this.props;

    if (isOpen) {
      if (!isModal && isDraggable) {
        const draggableProps = {};
        if (cancelDraggable) {
          draggableProps.cancel = cancelDraggable;
        }
        if (onDragStop) {
          draggableProps.onStop = evt => onDragStop(evt);
        }
        if (position) {
          draggableProps.position = position;
        }

        return <Draggable {...draggableProps}>{this.renderDialog()}</Draggable>;
      }
      return <>{this.renderDialog()}</>;
    }
    return null;
  }
}

Dialog.propTypes = propTypes;
Dialog.defaultProps = defaultProps;

export default Dialog;
