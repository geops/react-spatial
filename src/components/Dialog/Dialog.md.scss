.tm-dialog-example {
  position: relative;
  height: 200px;
  justify-content: center;
  align-items: center;

  .tm-dialog-content {
    padding: 15px;
  }

  .tm-dialog-btn {
    position: absolute;
    bottom: 0;
    margin-right: 15px;
    width: 150px;
    padding: 5px;
    border-radius: 10px;
    background-color: white;
    border: 2px solid $gray-lighter;

    &:hover {
      background-color: $gray-lighter;
    }
  }
}
