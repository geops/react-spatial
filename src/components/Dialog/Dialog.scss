.tm-dialog-container {
  background: white;
  text-align: center;
  width: 60%;

  .tm-dialog-header {
    position: relative;
    background: white;
    text-align: center;
    line-height: $btn-size-base;
    height: $btn-size-base;
  }

  .tm-dialog-draggable {
    cursor: pointer;
  }

  .tm-dialog-close-bt {
    position: absolute;
    border: none;
    right: 0;
    top: 0;
    color: black;
    height: $btn-size-base;
    width: $btn-size-base;
  }
}

.tm-modal {
  width: 100%;
  height: 100%;
  position: fixed;
  z-index: 2000;
  left: 0;
  top: 0;
  background: rgba(0, 0, 0, 0.1);

  .tm-dialog-container {
    position: absolute;
    height: auto;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  }
}

.tm-dialog {
  .tm-dialog-container {
    box-shadow: $box-shadow;
    margin: auto;
    position: absolute;
    top: 0;
  }
}
