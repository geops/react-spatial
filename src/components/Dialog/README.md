#

This demonstrates the use of Dialog.

```jsx
import React from  'react'
import Button from 'react-spatial/components/Button';
import Dialog from 'react-spatial/components/Dialog';

class DialogExample extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      isOpen: false,
    };
  }

  toggleDialog() {
    const { isOpen } = this.state;
    this.setState({
      isOpen: !isOpen,
    });
  }

  render() {
    const { isOpen } = this.state;
    return (
      <div className="tm-dialog-example">
        <Dialog
          title={<span>Example Dialog</span>}
          classNameChildren="tm-dialog-content"
          onClose={() => this.toggleDialog()}
          isOpen={isOpen}
          isDraggable
        >
          <span >I am the content of the dialog</span>
        </Dialog>
        <Button
          className="tm-button tm-dialog-btn"
          onClick={() => this.toggleDialog()}
        >
          Open Dialog
        </Button>
      </div>
    );
  }
}

<DialogExample />;
```
