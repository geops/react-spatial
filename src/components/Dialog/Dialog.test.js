import React from 'react';
import renderer from 'react-test-renderer';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Dialog from '.';

configure({ adapter: new Adapter() });

describe('Dialog', () => {
  describe('matches snapshots', () => {
    test('for closed Dialog.', () => {
      const component = renderer.create(
        <Dialog onClose={() => {}} isOpen={false} />,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('for normal Dialog.', () => {
      const component = renderer.create(
        <Dialog
          title={<span>Example Dialog</span>}
          onClose={() => {}}
          isDraggable
          isOpen
        >
          <span>content</span>
        </Dialog>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('for modal Dialog.', () => {
      const component = renderer.create(
        <Dialog onClose={() => {}} isModal isOpen>
          <span>content</span>
        </Dialog>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  test('triggers onClose.', () => {
    const fn = jest.fn();
    const wrapper = mount(
      <Dialog onClose={fn} isOpen classNameCloseBt="tm-close-classname">
        <span>content</span>
      </Dialog>,
    );

    wrapper
      .find('.tm-close-classname')
      .at(1)
      .simulate('click');
    expect(fn).toHaveBeenCalledTimes(1);
  });

  test('set dialog position.', () => {
    const fn = jest.fn();
    const wrapper = mount(
      <Dialog onClose={fn} isOpen isDraggable>
        <span>content</span>
      </Dialog>,
    );
    wrapper.setProps({
      position: {
        x: 9,
        y: 9,
      },
    });

    const draggable = wrapper.find('.react-draggable');
    expect(draggable.prop('style').transform).toBe('translate(9px,9px)');
  });
});
