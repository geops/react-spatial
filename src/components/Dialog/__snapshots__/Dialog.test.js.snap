// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog matches snapshots for closed Dialog. 1`] = `null`;

exports[`Dialog matches snapshots for modal Dialog. 1`] = `
<div
  className="tm-modal"
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    className="tm-dialog-container"
  >
    <div
      className="tm-dialog-header "
    >
      <div
        className="tm-header-left"
      />
      <div
        className="tm-button tm-dialog-close-bt"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <svg
          fill="currentColor"
          focusable={false}
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
    </div>
    <div>
      <span>
        content
      </span>
    </div>
  </div>
</div>
`;

exports[`Dialog matches snapshots for normal Dialog. 1`] = `
<div
  className="tm-dialog react-draggable"
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "display": "block",
      "touchAction": "none",
      "transform": "translate(0px,0px)",
    }
  }
  transform={null}
>
  <div
    className="tm-dialog-container"
  >
    <div
      className="tm-dialog-header tm-dialog-draggable"
    >
      <div
        className="tm-header-left"
      />
      <span>
        Example Dialog
      </span>
      <div
        className="tm-button tm-dialog-close-bt"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <svg
          fill="currentColor"
          focusable={false}
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
    </div>
    <div>
      <span>
        content
      </span>
    </div>
  </div>
</div>
`;
